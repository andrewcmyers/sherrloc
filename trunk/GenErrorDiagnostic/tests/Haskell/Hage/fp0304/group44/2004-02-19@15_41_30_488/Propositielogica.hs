module Propositielogica where

type Bedeling = [String]

elemBy        :: (a -> a -> Bool) -> a -> [a] -> Bool
elemBy = undefined
eqString      :: String -> String -> Bool 
eqString = undefined
eqList        :: (a -> a -> Bool) -> [a] -> [a] -> Bool
eqList = undefined
eqChar      :: Char -> Char -> Bool
eqChar = undefined
eqBool      :: Bool -> Bool -> Bool 
eqBool = undefined

data Prop
  = En [Prop]
  | Of [Prop]
  | Niet Prop
  | Prop :-> Prop
  | Var String
  | Bool Bool

evalueer :: Prop -> Bedeling -> Bool
evalueer (Bool x) _ = x
evalueer (Var p) lijst = elemBy eqString p lijst
evalueer (Niet p) bedeling = not(evalueer p bedeling)
evalueer (Of [] ) _ = False
evalueer (Of (x:xs)) bedeling = evalueer x bedeling || evalueer (Of xs) bedeling
evalueer (En [] ) _ = True
evalueer (En (x:xs)) bedeling = evalueer x bedeling && evalueer (En xs) bedeling
evalueer (p :-> q) bedeling        |evalueer p bedeling && evalueer q bedeling = True
                                   |evalueer (Niet p) bedeling = True
                                   |evalueer p bedeling && evalueer (Niet q) bedeling = False
                                   |otherwise = True


subs :: Bedeling -> [Bedeling]
subs [] = [[]]
subs  (x:xs) = map (x:) (subs xs) ++ subs xs


maakAlles :: Prop -> [String]
maakAlles (Bool _)  = []
maakAlles (Var p)   = [p]
maakAlles (En (x:xs)) = maakAlles x ++ maakAlles (En xs)
maakAlles (En [])   = []
maakAlles (Of prop)   = maakAlles (En prop)
maakAlles (Niet prop) = maakAlles prop
maakAlles (p :-> q)   = maakAlles p ++ maakAlles q


removeDupli :: [String] -> [String]
removeDupli [] = []
removeDupli (s:ss) |elemBy eqString s ss = removeDupli ss
                   |otherwise      = s: removeDupli ss


maakAllesEnSchoon :: Prop -> Bedeling
maakAllesEnSchoon prop = removeDupli (maakAlles prop)



vervulbaar :: Prop -> [Bedeling]
vervulbaar prop = filter (evalueer prop) ((subs.maakAllesEnSchoon) prop)


tautologie :: Prop -> Bool
tautologie prop    |vergelijkLijstBedeling (vervulbaar prop)(subs(maakAllesEnSchoon prop)) = True
                   |otherwise = False



vergelijkLijstBedeling :: [Bedeling] -> [Bedeling] -> Bool
vergelijkLijstBedeling (x:xs) volleProp            |length(filter(eqBool True)(map (eqList x) volleProp)) == 1 = True
                                                   |otherwise = False

-- should be (eqList eqString x)
-- 75,85-90   75,92-92   75,85-92
