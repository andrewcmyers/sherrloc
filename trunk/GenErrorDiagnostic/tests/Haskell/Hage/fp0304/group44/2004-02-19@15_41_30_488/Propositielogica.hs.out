Compiling Propositielogica.hs
(75,85)-(75,91): Type error in application
 expression       : eqList x
 term             : eqList
   type           : (a -> a -> Bool) -> [a]      -> [a] -> Bool
   does not match : Bedeling         -> Bedeling -> Bool       
 probable fix     : insert a first argument

Compilation failed with 1 error
